src/sentry/taskworker/consumer_grpc_push.py

```diff
@@ -18,29 +23,51 @@
 
 
 class ConsumerGrpc:
-    def __init__(self) -> None:
+    def __init__(self, worker_addrs: Iterable[str]) -> None:
         self.pending_task_store = PendingTaskStore()
-        self.host = "localhost"
-        self.server_port = 50051
-        self.channel = grpc.insecure_channel(f"{self.host}:{self.server_port}")
-        self.stub = WorkerServiceStub(self.channel)
+        self.available_stubs = deque(
+            [WorkerServiceStub(grpc.insecure_channel(worker_addr)) for worker_addr in worker_addrs]
+        )
+        self.current_connections = set()
 
     def start(self):
-        while True:
-            self.dispatch_task()
+        with ThreadPoolExecutor(max_workers=len(self.available_stubs)) as executor:
+            logger.info("Starting consumer grpc with %s threads", len(self.available_stubs))
+            while True:
+                inflight_activation = self._poll_pending_task()
+
+                if len(self.available_stubs) == 0:
+                    done, not_done = wait(self.current_connections, return_when=FIRST_COMPLETED)
+                    self.available_stubs.extend([future.result() for future in done])
+                    self.current_connections = not_done
+
+                self.current_connections.add(
+                    executor.submit(
+                        self._dispatch_activation,
+                        self.available_stubs.popleft(),
+                        inflight_activation,
+                    )
+                )
 
-    def dispatch_task(self):
-        in_flight_activation = self.pending_task_store.get_pending_task()
-        if not in_flight_activation:
+    def _poll_pending_task(self) -> InflightActivation:
+        while True:
+            inflight_activation = self.pending_task_store.get_pending_task()
+            if inflight_activation:
+                return inflight_activation
             logger.info("No tasks")
             time.sleep(1)
```

Comment:
We're still polling here, not sure if there's a way to have django wake us up instead of hitting the db constantly

Reply:
Just double checking my understanding here. Do we know if this polling happens in the main thread or in the child thread(s)? Looking into the code, it looks like this happens in the main since we're executing `_poll_pending_task()` and passing the return value into the executor? I think that would be decent implementation (poll inflight activation, check if we have available stub, spawn thread to work on the task). That way we don't have these long running threads that poll.

Reply:
You're right, the polling happens only in the main thread. The child threads only starts when a connection is available